-code-checks

ERROR: Build errors found during clang-tidy run.
```
gmake: *** [tmp/slc7_amd64_gcc820/Makefile:32: .SCRAM/slc7_amd64_gcc820/timestamps/coral] Error 1
gzip: /build/cmsbld/jenkins/workspace/run-pr-code-checks/CMSSW_11_2_X_2020-07-08-2300/.SCRAM/slc7_amd64_gcc820/ToolCache.db.gz: No such file or directory
Can't call method "initpathvars" on an undefined value at /cvmfs/cms-ib.cern.ch/nweek-02636/share/lcg/SCRAMV1/V2_2_9_pre07/src/SCRAM/SCRAM.pm line 392.
	slc7_amd64_gcc820
--
gmake: *** [tmp/slc7_amd64_gcc820/Makefile:32: .SCRAM/slc7_amd64_gcc820/timestamps/cmsswdata] Error 1
gmake: *** [tmp/slc7_amd64_gcc820/Makefile:32: .SCRAM/slc7_amd64_gcc820/timestamps/codechecker] Error 1
gmake: *** [tmp/slc7_amd64_gcc820/Makefile:32: .SCRAM/slc7_amd64_gcc820/timestamps/csctrackfinderemulation] Error 1
You are trying to build/setup tools for SCRAM_ARCH slc7_amd64_gcc820.
while your project area is currently setup for following SCRAM_ARCH(s):
	slc7_amd64_gcc820
--
gmake: *** [tmp/slc7_amd64_gcc820/Makefile:32: .SCRAM/slc7_amd64_gcc820/timestamps/clhepheader] Error 1

gzip: /build/cmsbld/jenkins/workspace/run-pr-code-checks/CMSSW_11_2_X_2020-07-08-2300/.SCRAM/slc7_amd64_gcc820/ToolCache.db.gz: unexpected end of file
Can't call method "initpathvars" on an undefined value at /cvmfs/cms-ib.cern.ch/nweek-02636/share/lcg/SCRAMV1/V2_2_9_pre07/src/SCRAM/SCRAM.pm line 392.
gmake: *** [tmp/slc7_amd64_gcc820/Makefile:32: .SCRAM/slc7_amd64_gcc820/timestamps/cppunit] Error 1
gmake: *** [tmp/slc7_amd64_gcc820/Makefile:32: .SCRAM/slc7_amd64_gcc820/timestamps/cuda] Error 1
Updating symlinks under external/slc7_amd64_gcc820

Finding a value for CLHEP_BASE:
--
gmake: *** No rule to make target 'code-checks'.
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

--- a/Validation/MuonCSCDigis/src/CSCDigiMatcher.cc
+++ b/Validation/MuonCSCDigis/src/CSCDigiMatcher.cc
@@ -57,16 +57,6 @@ void CSCDigiMatcher::match(const SimTrack& t, const SimVertex& v) {
 }
 
 void CSCDigiMatcher::matchComparatorsToSimTrack(const CSCComparatorDigiCollection& comparators) {
-  for (auto detUnitIt = comparators.begin(); detUnitIt != comparators.end(); ++detUnitIt) {
-    const CSCDetId& id = (*detUnitIt).first;
-    const auto& range = (*detUnitIt).second;
-    for (auto digiIt = range.first; digiIt != range.second; ++digiIt) {
-      if (id.station() == 1 and (id.ring() == 1 or id.ring() == 4))
-        if (verboseComparator_)
-          cout << "CSCid " << id << " Comparator digi (comparator, comparator, Tbin ) " << (*digiIt) << endl;
-    }
-  }
-
   const auto& det_ids = muonSimHitMatcher_->detIds(0);
   for (const auto& id : det_ids) {
     CSCDetId layer_id(id);

--- a/Validation/MuonCSCDigis/src/CSCStubMatcher.cc
+++ b/Validation/MuonCSCDigis/src/CSCStubMatcher.cc
@@ -1,6 +1,5 @@
 #include "Validation/MuonCSCDigis/interface/CSCStubMatcher.h"
 #include "L1Trigger/CSCCommonTrigger/interface/CSCConstants.h"
-
 #include <algorithm>
 
 using namespace std;
@@ -87,12 +86,9 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
     }
 
     int ring = ch_id.ring();
-    if (ring == 4)
-      ring = 1;  //use ME1b id to get CLCTs
-    CSCDetId ch_id2(ch_id.endcap(), ch_id.station(), ring, ch_id.chamber(), 0);
 
     // do not consider CSCs with too few hits
-    if (cscDigiMatcher_->nLayersWithStripInChamber(ch_id2) < minNHitsChamberCLCT_)
+    if (cscDigiMatcher_->nLayersWithStripInChamber(ch_id) < minNHitsChamberCLCT_)
       continue;
 
     // get the comparator digis in this chamber
@@ -104,7 +100,7 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
 
     // print out the digis
     if (verboseCLCT_) {
-      cout << "clct: digi_strips " << ch_id2 << endl;
+      cout << "clct: comparators " << ch_id << endl;
       int layer = 0;
       for (const auto& p : comps) {
         layer++;
@@ -114,6 +110,12 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
         cout << endl;
       }
     }
+
+    //use ME1b id to get CLCTs
+    const bool isME1a(ch_id.station() == 1 and ch_id.ring() == 4);
+    if (isME1a)
+      ring = 1;
+    CSCDetId ch_id2(ch_id.endcap(), ch_id.station(), ring, ch_id.chamber(), 0);
 
     const auto& clcts_in_det = clcts.get(ch_id2);
 
@@ -137,15 +139,27 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
       for (const auto& p : comps) {
         layer++;
         for (const auto& q : p) {
+          if (verboseCLCT_)
+            cout << "L" << layer << " " << q << " " << q.getHalfStrip() << " " << std::endl;
           for (const auto& clctComp : (*c).getHits()[layer - 1]) {
-            if (q.getHalfStrip() == clctComp) {
+            if (clctComp == 65535) continue;
+            if (verboseCLCT_) {
+              std::cout << "\t" << clctComp << " " << endl;
+            }
+            if (q.getHalfStrip() == clctComp or (isME1a and q.getHalfStrip() + 128 == clctComp)) {
               nMatches++;
+              if (verboseCLCT_) {
+                cout << "\t\tnMatches " << nMatches << std::endl;
+              }
             }
           }
         }
+        if (verboseCLCT_) {
+          cout << endl;
+        }
       }
 
-      // require at least 3 good half-strip matches
+      // require at least 3 good matches
       if (nMatches < 3)
         continue;
 
@@ -257,19 +271,26 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
     for (const auto& lct : lcts_tmp) {
       iLct++;
 
-      bool lct_matched(false);
       bool lct_clct_match(false);
       bool lct_alct_match(false);
       bool lct_gem1_match(false);
       bool lct_gem2_match(false);
 
-      if (verboseLCT_)
-        cout << "in LCT, getCLCT " << lct.getCLCT() << " getALCT " << lct.getALCT() << endl;
-
+      if (verboseLCT_) {
+        cout << ch_id << " " << ch_id2 << endl;
+        cout << lct << endl;
+        cout << "getCLCT " << lct.getCLCT()
+             << "\ngetALCT " << lct.getALCT()
+             << "\ngetGEM1 " << lct.getGEM1()
+             << "\ngetGEM2 " << lct.getGEM2()
+             << endl;
+      }
       // Check if matched to an CLCT
       for (const auto& p : clctsInChamber(id)) {
         if (p == lct.getCLCT()) {
           lct_clct_match = true;
+          if (verboseLCT_)
+            cout << "\t...lct_clct_match" << endl;
           break;
         }
       }
@@ -278,6 +299,8 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
       for (const auto& p : alctsInChamber(id)) {
         if (p == lct.getALCT()) {
           lct_alct_match = true;
+          if (verboseLCT_)
+            cout << "\t...lct_alct_match" << endl;
           break;
         }
       }
@@ -289,26 +312,35 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
         for (const auto& p : gemDigiMatcher_->padsInChamber(gemDetIdL1.rawId())) {
           if (p == lct.getGEM1()) {
             lct_gem1_match = true;
+            if (verboseLCT_)
+              cout << "\t...lct_gem1_match" << endl;
             break;
-          }
+            }
         }
 
         // Check if matched to an GEM pad L2
-        const GEMDetId gemDetIdL2(ch_id.zendcap(), 1, ch_id.station(), 2, ch_id.chamber(), 0);
-        for (const auto& p : gemDigiMatcher_->padsInChamber(gemDetIdL2.rawId())) {
+          const GEMDetId gemDetIdL2(ch_id.zendcap(), 1, ch_id.station(), 2, ch_id.chamber(), 0);
+          for (const auto& p : gemDigiMatcher_->padsInChamber(gemDetIdL2.rawId())) {
           if (p == lct.getGEM2()) {
             lct_gem2_match = true;
+            if (verboseLCT_)
+              cout << "\t...lct_gem2_match" << endl;
             break;
           }
         }
       }
 
-      lct_matched = ((lct_clct_match and lct_alct_match) or (lct_alct_match and lct_gem1_match and lct_gem2_match) or
-                     (lct_clct_match and lct_gem1_match and lct_gem2_match));
+      const bool alct_clct = lct_clct_match and lct_alct_match;
+      const bool alct_gem = lct_alct_match and lct_gem1_match and lct_gem2_match;
+      const bool clct_gem = lct_clct_match and lct_gem1_match and lct_gem2_match;
+
+      bool lct_tight_matched = alct_clct or alct_gem or clct_gem;
+      bool lct_loose_matched = lct_clct_match or lct_alct_match;
+      bool lct_matched = lct_loose_matched or lct_tight_matched;
 
       if (lct_matched) {
         if (verboseLCT_)
-          cout << "this LCT matched to simtrack in chamber " << ch_id << endl;
+          cout << "...was matched" << endl;
         if (std::find(chamber_to_lcts_[id].begin(), chamber_to_lcts_[id].end(), lct) == chamber_to_lcts_[id].end()) {
           chamber_to_lcts_[id].emplace_back(lct);
         }
@@ -331,7 +363,6 @@ void CSCStubMatcher::matchMPLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection&
       if (!lct->isValid())
         continue;
 
-      // std::cout << "MPC Stub ALL " << *lct << std::endl;
       chamber_to_mplcts_all_[id].emplace_back(*lct);
 
       // check if this stub corresponds with a previously matched stub

--- a/Validation/MuonCSCDigis/src/CSCStubMatcher.cc
+++ b/Validation/MuonCSCDigis/src/CSCStubMatcher.cc
@@ -142,7 +142,8 @@ void CSCStubMatcher::matchCLCTsToSimTrack(const CSCCLCTDigiCollection& clcts) {
           if (verboseCLCT_)
             cout << "L" << layer << " " << q << " " << q.getHalfStrip() << " " << std::endl;
           for (const auto& clctComp : (*c).getHits()[layer - 1]) {
-            if (clctComp == 65535) continue;
+            if (clctComp == 65535)
+              continue;
             if (verboseCLCT_) {
               std::cout << "\t" << clctComp << " " << endl;
             }
@@ -279,11 +280,8 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
       if (verboseLCT_) {
         cout << ch_id << " " << ch_id2 << endl;
         cout << lct << endl;
-        cout << "getCLCT " << lct.getCLCT()
-             << "\ngetALCT " << lct.getALCT()
-             << "\ngetGEM1 " << lct.getGEM1()
-             << "\ngetGEM2 " << lct.getGEM2()
-             << endl;
+        cout << "getCLCT " << lct.getCLCT() << "\ngetALCT " << lct.getALCT() << "\ngetGEM1 " << lct.getGEM1()
+             << "\ngetGEM2 " << lct.getGEM2() << endl;
       }
       // Check if matched to an CLCT
       for (const auto& p : clctsInChamber(id)) {
@@ -315,12 +313,12 @@ void CSCStubMatcher::matchLCTsToSimTrack(const CSCCorrelatedLCTDigiCollection& l
             if (verboseLCT_)
               cout << "\t...lct_gem1_match" << endl;
             break;
-            }
+          }
         }
 
         // Check if matched to an GEM pad L2
-          const GEMDetId gemDetIdL2(ch_id.zendcap(), 1, ch_id.station(), 2, ch_id.chamber(), 0);
-          for (const auto& p : gemDigiMatcher_->padsInChamber(gemDetIdL2.rawId())) {
+        const GEMDetId gemDetIdL2(ch_id.zendcap(), 1, ch_id.station(), 2, ch_id.chamber(), 0);
+        for (const auto& p : gemDigiMatcher_->padsInChamber(gemDetIdL2.rawId())) {
           if (p == lct.getGEM2()) {
             lct_gem2_match = true;
             if (verboseLCT_)

--- a/Validation/MuonGEMDigis/interface/GEMDigiMatcher.h
+++ b/Validation/MuonGEMDigis/interface/GEMDigiMatcher.h
@@ -78,6 +78,7 @@ public:
   // #layers with digis from this simtrack
   int nLayersWithDigisInSuperChamber(unsigned int) const;
   int nLayersWithPadsInSuperChamber(unsigned int) const;
+  int nLayersWithClustersInSuperChamber(unsigned int) const;
 
   /// How many pads in GEM did this simtrack get in total?
   int nPads() const;

--- a/Validation/MuonGEMDigis/python/muonGEMDigiPSet.py
+++ b/Validation/MuonGEMDigis/python/muonGEMDigiPSet.py
@@ -30,7 +30,7 @@ muonGEMDigiPSet = cms.PSet(
     gemCoPadDigi = cms.PSet(
         verbose = cms.int32(0),
         inputTag = cms.InputTag("simCscTriggerPrimitiveDigis"),
-        minBX = cms.int32(-1),
-        maxBX = cms.int32(1),
+        minBX = cms.int32(0),
+        maxBX = cms.int32(0),
     ),
 )

--- a/Validation/MuonGEMDigis/src/GEMDigiMatcher.cc
+++ b/Validation/MuonGEMDigis/src/GEMDigiMatcher.cc
@@ -141,7 +141,6 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection& 
     auto clusters_in_det = clusters.get(p_id);
 
     for (auto cluster = clusters_in_det.first; cluster != clusters_in_det.second; ++cluster) {
-
       bool isMatched;
 
       if (verbosePad_)
@@ -173,15 +172,12 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection& 
 }
 
 void GEMDigiMatcher::matchCoPadsToSimTrack(const GEMCoPadDigiCollection& co_pads) {
-
   // loop on the GEM detids
   for (auto detUnitIt = co_pads.begin(); detUnitIt != co_pads.end(); ++detUnitIt) {
-
     const GEMDetId& id = (*detUnitIt).first;
     const auto& co_pads_in_det = (*detUnitIt).second;
 
     for (auto copad = co_pads_in_det.first; copad != co_pads_in_det.second; ++copad) {
-
       // check that the cluster BX is within the range
       if (copad->bx(1) < minBXCoPad_ || copad->bx(1) > maxBXCoPad_)
         continue;
@@ -195,14 +191,14 @@ void GEMDigiMatcher::matchCoPadsToSimTrack(const GEMCoPadDigiCollection& co_pads
         cout << "GEMCoPadDigi: " << id << " " << *copad << endl;
 
       // first pad is tightly matched
-      for (const auto& p: padsInDetId(gemL1_id.rawId())) {
+      for (const auto& p : padsInDetId(gemL1_id.rawId())) {
         if (p == copad->first()) {
           isMatchedL1 = true;
         }
       }
 
       // second pad can only be loosely matched
-      for (const auto& p: padsInChamber(gemL2_id.rawId())) {
+      for (const auto& p : padsInChamber(gemL2_id.rawId())) {
         if (p == copad->second()) {
           isMatchedL2 = true;
         }

--- a/Validation/MuonGEMDigis/src/GEMDigiMatcher.cc
+++ b/Validation/MuonGEMDigis/src/GEMDigiMatcher.cc
@@ -85,21 +85,24 @@ void GEMDigiMatcher::matchDigisToSimTrack(const GEMDigiCollection& digis) {
     const auto& digis_in_det = digis.get(GEMDetId(id));
 
     for (auto d = digis_in_det.first; d != digis_in_det.second; ++d) {
+
       if (verboseDigi_)
         cout << "GEMDigi " << p_id << " " << *d << endl;
+
       // check that the digi is within BX range
       if (d->bx() < minBXDigi_ || d->bx() > maxBXDigi_)
         continue;
+
       // check that it matches a strip that was hit by SimHits from our track
       if (hit_strips.find(d->strip()) == hit_strips.end())
         continue;
+
       if (verboseDigi_)
         cout << "...was matched!" << endl;
 
       detid_to_digis_[id].push_back(*d);
       chamber_to_digis_[p_id.chamberId().rawId()].push_back(*d);
       superchamber_to_digis_[p_id.superChamberId().rawId()].push_back(*d);
-      break;
     }
   }
 }
@@ -112,21 +115,29 @@ void GEMDigiMatcher::matchPadsToSimTrack(const GEMPadDigiCollection& pads) {
     const auto& pads_in_det = pads.get(p_id);
 
     for (auto pad = pads_in_det.first; pad != pads_in_det.second; ++pad) {
+
+      bool isMatched;
+
       if (verbosePad_)
         cout << "GEMPad " << p_id << " " << *pad << endl;
+
       // check that the pad BX is within the range
       if (pad->bx() < minBXPad_ || pad->bx() > maxBXPad_)
         continue;
+
       // check that it matches a pad that was hit by SimHits from our track
       for (auto digi : detid_to_digis_[id]) {
         if (digi.strip() / 2 == pad->pad()) {
-          detid_to_pads_[id].push_back(*pad);
-          chamber_to_pads_[p_id.chamberId().rawId()].push_back(*pad);
-          superchamber_to_pads_[p_id.superChamberId().rawId()].push_back(*pad);
-          if (verbosePad_)
-            cout << "...was matched!" << endl;
+          isMatched = true;
           break;
         }
+      }
+      if (isMatched) {
+        detid_to_pads_[p_id.rawId()].push_back(*pad);
+        chamber_to_pads_[p_id.chamberId().rawId()].push_back(*pad);
+        superchamber_to_pads_[p_id.superChamberId().rawId()].push_back(*pad);
+        if (verbosePad_)
+          cout << "...was matched!" << endl;
       }
     }
   }
@@ -136,7 +147,6 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection& 
   const auto& det_ids = muonSimHitMatcher_->detIds();
   for (auto id : det_ids) {
     GEMDetId p_id(id);
-    GEMDetId superch_id(p_id.region(), p_id.ring(), p_id.station(), 1, p_id.chamber(), 0);
 
     auto clusters_in_det = clusters.get(p_id);
 
@@ -162,10 +172,9 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection& 
       if (isMatched) {
         detid_to_clusters_[id].push_back(*cluster);
         chamber_to_clusters_[p_id.chamberId().rawId()].push_back(*cluster);
-        superchamber_to_clusters_[superch_id()].push_back(*cluster);
+        superchamber_to_clusters_[p_id.superChamberId().rawId()].push_back(*cluster);
         if (verboseCluster_)
           cout << "...was matched!" << endl;
-        break;
       }
     }
   }
@@ -178,6 +187,10 @@ void GEMDigiMatcher::matchCoPadsToSimTrack(const GEMCoPadDigiCollection& co_pads
     const auto& co_pads_in_det = (*detUnitIt).second;
 
     for (auto copad = co_pads_in_det.first; copad != co_pads_in_det.second; ++copad) {
+
+      if (verboseCoPad_)
+        cout << "GEMCoPadDigi: " << id << " " << *copad << endl;
+
       // check that the cluster BX is within the range
       if (copad->bx(1) < minBXCoPad_ || copad->bx(1) > maxBXCoPad_)
         continue;
@@ -186,9 +199,6 @@ void GEMDigiMatcher::matchCoPadsToSimTrack(const GEMCoPadDigiCollection& co_pads
       bool isMatchedL2 = false;
       GEMDetId gemL1_id(id.region(), 1, id.station(), 1, id.chamber(), copad->roll());
       GEMDetId gemL2_id(id.region(), 1, id.station(), 2, id.chamber(), 0);
-
-      if (verboseCoPad_)
-        cout << "GEMCoPadDigi: " << id << " " << *copad << endl;
 
       // first pad is tightly matched
       for (const auto& p : padsInDetId(gemL1_id.rawId())) {
@@ -204,7 +214,7 @@ void GEMDigiMatcher::matchCoPadsToSimTrack(const GEMCoPadDigiCollection& co_pads
         }
       }
       if (isMatchedL1 and isMatchedL2) {
-        superchamber_to_copads_[id].push_back(*copad);
+        superchamber_to_copads_[id.rawId()].push_back(*copad);
         if (verboseCoPad_)
           cout << "...was matched! " << endl;
       }

--- a/Validation/MuonGEMDigis/src/GEMDigiMatcher.cc
+++ b/Validation/MuonGEMDigis/src/GEMDigiMatcher.cc
@@ -99,6 +99,7 @@ void GEMDigiMatcher::matchDigisToSimTrack(const GEMDigiCollection& digis) {
       detid_to_digis_[id].push_back(*d);
       chamber_to_digis_[p_id.chamberId().rawId()].push_back(*d);
       superchamber_to_digis_[p_id.superChamberId().rawId()].push_back(*d);
+      break;
     }
   }
 }
@@ -107,35 +108,26 @@ void GEMDigiMatcher::matchPadsToSimTrack(const GEMPadDigiCollection& pads) {
   const auto& det_ids = muonSimHitMatcher_->detIds();
   for (const auto& id : det_ids) {
     GEMDetId p_id(id);
-    GEMDetId superch_id(p_id.region(), p_id.ring(), p_id.station(), 0, p_id.chamber(), 0);
 
-    const auto& hit_pads = muonSimHitMatcher_->hitPadsInDetId(id);
     const auto& pads_in_det = pads.get(p_id);
-
-    if (verbosePad_) {
-      cout << "checkpads " << hit_pads.size() << " " << std::distance(pads_in_det.first, pads_in_det.second)
-           << " hit_pads: ";
-      copy(hit_pads.begin(), hit_pads.end(), ostream_iterator<int>(cout, " "));
-      cout << endl;
-    }
 
     for (auto pad = pads_in_det.first; pad != pads_in_det.second; ++pad) {
       if (verbosePad_)
-        cout << "chp " << *pad << endl;
+        cout << "GEMPad " << p_id << " " << *pad << endl;
       // check that the pad BX is within the range
       if (pad->bx() < minBXPad_ || pad->bx() > maxBXPad_)
         continue;
-      if (verbosePad_)
-        cout << "chp1" << endl;
       // check that it matches a pad that was hit by SimHits from our track
-      if (hit_pads.find(pad->pad()) == hit_pads.end())
-        continue;
-      if (verbosePad_)
-        cout << "chp2" << endl;
-
-      detid_to_pads_[id].push_back(*pad);
-      chamber_to_pads_[p_id.chamberId().rawId()].push_back(*pad);
-      superchamber_to_pads_[superch_id()].push_back(*pad);
+      for (auto digi : detid_to_digis_[id]) {
+        if (digi.strip() / 2 == pad->pad()) {
+          detid_to_pads_[id].push_back(*pad);
+          chamber_to_pads_[p_id.chamberId().rawId()].push_back(*pad);
+          superchamber_to_pads_[p_id.superChamberId().rawId()].push_back(*pad);
+          if (verbosePad_)
+            cout << "...was matched!" << endl;
+          break;
+        }
+      }
     }
   }
 }
@@ -146,70 +138,79 @@ void GEMDigiMatcher::matchClustersToSimTrack(const GEMPadDigiClusterCollection& 
     GEMDetId p_id(id);
     GEMDetId superch_id(p_id.region(), p_id.ring(), p_id.station(), 1, p_id.chamber(), 0);
 
-    auto hit_pads = muonSimHitMatcher_->hitPadsInDetId(id);
     auto clusters_in_det = clusters.get(p_id);
 
     for (auto cluster = clusters_in_det.first; cluster != clusters_in_det.second; ++cluster) {
+
+      bool isMatched;
+
+      if (verbosePad_)
+        cout << "GEMCluster " << p_id << " " << *cluster << endl;
+
       // check that the cluster BX is within the range
       if (cluster->bx() < minBXCluster_ || cluster->bx() > maxBXCluster_)
         continue;
 
       // check that at least one pad was hit by the track
       for (const auto& p : cluster->pads()) {
-        if (hit_pads.find(p) == hit_pads.end()) {
-          detid_to_clusters_[id].push_back(*cluster);
-          chamber_to_clusters_[p_id.chamberId().rawId()].push_back(*cluster);
-          superchamber_to_clusters_[superch_id()].push_back(*cluster);
-          break;
+        for (auto pad : detid_to_pads_[id]) {
+          if (pad.pad() == p) {
+            isMatched = true;
+            break;
+          }
         }
+      }
+      if (isMatched) {
+        detid_to_clusters_[id].push_back(*cluster);
+        chamber_to_clusters_[p_id.chamberId().rawId()].push_back(*cluster);
+        superchamber_to_clusters_[superch_id()].push_back(*cluster);
+        if (verboseCluster_)
+          cout << "...was matched!" << endl;
+        break;
       }
     }
   }
 }
 
 void GEMDigiMatcher::matchCoPadsToSimTrack(const GEMCoPadDigiCollection& co_pads) {
-  const auto& det_ids = muonSimHitMatcher_->detIdsCoincidences();
-  for (const auto& id : det_ids) {
-    GEMDetId p_id(id);
-    GEMDetId superch_id(p_id.region(), p_id.ring(), p_id.station(), 0, p_id.chamber(), 0);
 
-    const auto& hit_co_pads = muonSimHitMatcher_->hitCoPadsInDetId(id);
-    const auto& co_pads_in_det = co_pads.get(superch_id);
+  // loop on the GEM detids
+  for (auto detUnitIt = co_pads.begin(); detUnitIt != co_pads.end(); ++detUnitIt) {
 
-    if (verboseCoPad_) {
-      cout << "matching CoPads in detid " << superch_id << std::endl;
-      cout << "checkcopads from gemhits" << hit_co_pads.size() << " from copad collection "
-           << std::distance(co_pads_in_det.first, co_pads_in_det.second) << " hit_pads: ";
-      copy(hit_co_pads.begin(), hit_co_pads.end(), ostream_iterator<int>(cout, " "));
-      cout << endl;
-    }
+    const GEMDetId& id = (*detUnitIt).first;
+    const auto& co_pads_in_det = (*detUnitIt).second;
 
-    for (auto pad = co_pads_in_det.first; pad != co_pads_in_det.second; ++pad) {
-      // to match simtrack to GEMCoPad, check the pads within the copad!
-      bool matchL1 = false;
-      GEMDetId gemL1_id(p_id.region(), p_id.ring(), p_id.station(), 1, p_id.chamber(), 0);
+    for (auto copad = co_pads_in_det.first; copad != co_pads_in_det.second; ++copad) {
+
+      // check that the cluster BX is within the range
+      if (copad->bx(1) < minBXCoPad_ || copad->bx(1) > maxBXCoPad_)
+        continue;
+
+      bool isMatchedL1 = false;
+      bool isMatchedL2 = false;
+      GEMDetId gemL1_id(id.region(), 1, id.station(), 1, id.chamber(), copad->roll());
+      GEMDetId gemL2_id(id.region(), 1, id.station(), 2, id.chamber(), 0);
+
       if (verboseCoPad_)
-        cout << "CoPad: chp " << *pad << endl;
-      for (const auto& p : padsInChamber(gemL1_id.rawId())) {
-        if (p == pad->first()) {
-          matchL1 = true;
-          break;
+        cout << "GEMCoPadDigi: " << id << " " << *copad << endl;
+
+      // first pad is tightly matched
+      for (const auto& p: padsInDetId(gemL1_id.rawId())) {
+        if (p == copad->first()) {
+          isMatchedL1 = true;
         }
       }
 
-      bool matchL2 = false;
-      GEMDetId gemL2_id(p_id.region(), p_id.ring(), p_id.station(), 2, p_id.chamber(), 0);
-      for (const auto& p : padsInChamber(gemL2_id.rawId())) {
-        if (p == pad->second()) {
-          matchL2 = true;
-          break;
+      // second pad can only be loosely matched
+      for (const auto& p: padsInChamber(gemL2_id.rawId())) {
+        if (p == copad->second()) {
+          isMatchedL2 = true;
         }
       }
-
-      if (matchL1 and matchL2) {
+      if (isMatchedL1 and isMatchedL2) {
+        superchamber_to_copads_[id].push_back(*copad);
         if (verboseCoPad_)
-          cout << "CoPad: was matched! " << endl;
-        superchamber_to_copads_[superch_id()].push_back(*pad);
+          cout << "...was matched! " << endl;
       }
     }
   }
@@ -337,6 +338,21 @@ int GEMDigiMatcher::nLayersWithPadsInSuperChamber(unsigned int detid) const {
     const auto& pads = padsInChamber(ch_id.rawId());
     // at least one digi in this layer!
     if (!pads.empty()) {
+      layers.insert(iLayer);
+    }
+  }
+  return layers.size();
+}
+
+int GEMDigiMatcher::nLayersWithClustersInSuperChamber(unsigned int detid) const {
+  set<int> layers;
+  GEMDetId sch_id(detid);
+  for (int iLayer = 1; iLayer <= 2; iLayer++) {
+    GEMDetId ch_id(sch_id.region(), sch_id.ring(), sch_id.station(), iLayer, sch_id.chamber(), 0);
+    // get the pads in this chamber
+    const auto& clusters = clustersInChamber(ch_id.rawId());
+    // at least one digi in this layer!
+    if (!clusters.empty()) {
       layers.insert(iLayer);
     }
   }

--- a/Validation/MuonHits/interface/GEMSimHitMatcher.h
+++ b/Validation/MuonHits/interface/GEMSimHitMatcher.h
@@ -65,7 +65,7 @@ public:
   // calculate average strip number for a provided collection of simhits
   float simHitsMeanStrip(const edm::PSimHitContainer& sim_hits) const;
 
-  std::set<int> hitStripsInDetId(unsigned int, int margin_n_strips = 0) const;
+  std::set<int> hitStripsInDetId(unsigned int, int margin_n_strips = 1) const;
   std::set<int> hitPadsInDetId(unsigned int) const;
   std::set<int> hitCoPadsInDetId(unsigned int) const;
 

--- a/Validation/MuonHits/src/GEMSimHitMatcher.cc
+++ b/Validation/MuonHits/src/GEMSimHitMatcher.cc
@@ -95,7 +95,7 @@ void GEMSimHitMatcher::matchSimHitsToSimTrack() {
     set<int> pads;
     for (const auto& h : hits) {
       const LocalPoint& lp = h.entryPoint();
-      pads.insert(1 + static_cast<int>(roll->padTopology().channel(lp)));
+      pads.insert(static_cast<int>(roll->padTopology().channel(lp)));
     }
     detids_to_pads_[d] = pads;
   }
@@ -115,10 +115,10 @@ void GEMSimHitMatcher::matchSimHitsToSimTrack() {
 
     for (const auto& h : hits1) {
       const LocalPoint& lp = h.entryPoint();
-      pads1.insert(1 + static_cast<int>(roll1->padTopology().channel(lp)));
+      pads1.insert(static_cast<int>(roll1->padTopology().channel(lp)));
       if (verbose_)
         edm::LogInfo("GEMSimHitMatcher") << "GEMHits detid1 " << id1 << " pad1 "
-                                         << 1 + static_cast<int>(roll1->padTopology().channel(lp)) << std::endl;
+                                         << static_cast<int>(roll1->padTopology().channel(lp)) << std::endl;
     }
 
     // find pads with hits in layer2
@@ -133,10 +133,10 @@ void GEMSimHitMatcher::matchSimHitsToSimTrack() {
       const auto& roll2 = dynamic_cast<const GEMGeometry*>(geometry_)->etaPartition(id2);
       for (const auto& h : hits2) {
         const LocalPoint& lp = h.entryPoint();
-        pads2.insert(1 + static_cast<int>(roll2->padTopology().channel(lp)));
+        pads2.insert(static_cast<int>(roll2->padTopology().channel(lp)));
         if (verbose_)
           edm::LogInfo("GEMSimHitMatcher") << "GEMHits detid2 " << id2 << " pad2 "
-                                           << 1 + static_cast<int>(roll2->padTopology().channel(lp)) << std::endl;
+                                           << static_cast<int>(roll2->padTopology().channel(lp)) << std::endl;
       }
     }
 
@@ -296,7 +296,7 @@ std::set<int> GEMSimHitMatcher::hitStripsInDetId(unsigned int detid, int margin_
     int central_strip =
         static_cast<int>(dynamic_cast<const GEMGeometry*>(geometry_)->etaPartition(id)->topology().channel(lp));
     int smin = central_strip - margin_n_strips;
-    smin = (smin > 0) ? smin : 1;
+    smin = (smin > 0) ? smin : 0;
     int smax = central_strip + margin_n_strips;
     smax = (smax <= max_nstrips) ? smax : max_nstrips;
     for (int ss = smin; ss <= smax; ++ss)

